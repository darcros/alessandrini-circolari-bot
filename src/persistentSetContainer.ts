import { readFile, writeFile, mkdir } from "fs/promises";
import { dirname } from "path";

export class PersistentSetContainer<T> {
  private set: Set<T> = null;
  private readonly filePath: string;

  constructor(filePath: string) {
    this.filePath = filePath;
    this.load();
  }

  public async get(): Promise<Set<T>> {
    if (this.set === null) {
      this.set = await this.load();
    }

    return this.set;
  }

  private async load(): Promise<Set<T>> {
    try {
      const buf = await readFile(this.filePath);
      const arr = JSON.parse(buf.toString());
      return new Set(arr);
    } catch (_) {
      return new Set();
    }
  }

  public async save(): Promise<void> {
    const urls = [...this.set.values()];
    const jsonString = JSON.stringify(urls, null, 2);

    await mkdir(dirname(this.filePath), { recursive: true });
    await writeFile(this.filePath, jsonString);
  }
}
