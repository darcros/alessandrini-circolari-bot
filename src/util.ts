export function fixMultinePadding(s: string) {
  return s.split('\n')
    .map(s => s.trim())
    .join('\n');
}

export function requireEnv(name: string) {
  const envVar = process.env[name];

  if (!envVar) {
    const message = `
      VARIBILI D'AMBIENTE MANCANTI!
      Questa variabile d'ambiente non Ã¨ stata settata: ${name}.
      Puoi settarla in due modi:
      1. crea un file chiamato .env nella stessa cartella di package.json e aggiungi questa riga "${name}=valore"
      2. settala tramite il tuo sistema operativo (con bash, CMD, ecc..)
    `;

    throw new Error(fixMultinePadding(message));
  }

  return envVar;
}
