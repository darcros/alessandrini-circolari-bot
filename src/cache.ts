import { readFile, writeFile, mkdir, access } from 'fs/promises';
import { constants } from 'fs';
import { dirname } from 'path';
import { getEnv } from './util';

const CACHE_PATH = getEnv('CACHE_PATH', './data/cache.json');

function exists(path: string): Promise<boolean> {
  return access(path, constants.R_OK | constants.W_OK)
    .then(() => true)
    .catch(() => false);
}

async function read<T>(path: string): Promise<Set<T>> {
  const buf = await readFile(path);
  const arr = JSON.parse(buf.toString());
  return new Set(arr);
}

async function write<T>(path: string, set: Set<T>): Promise<void> {
  const urls = [...set.values()];
  const jsonString = JSON.stringify(urls, null, 2);

  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, jsonString);
}

export async function wasSent(url: string): Promise<boolean> {
  if (!exists(CACHE_PATH)) return false;

  const set = await read<string>(CACHE_PATH);
  return set.has(url);
}

export async function confirmSent(url: string): Promise<void> {
  const set = exists(CACHE_PATH) ? await read<string>(CACHE_PATH) : new Set();

  set.add(url);
  await write(CACHE_PATH, set);
}
