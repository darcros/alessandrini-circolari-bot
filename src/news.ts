import axios from 'axios';
import cheerio from 'cheerio';
import { parse } from 'date-fns';
import { URL } from 'url';
import { wasSent } from './cache';

function scrapeNewsList(page: string) {
  const $ = cheerio.load(page);

  const urls: string[] = [];

  $('.view-content > .views-row').each((_i, tag) => {
    const newsElement = $(tag);

    const titleElement = newsElement.find('.views-field-title > .field-content > a');
    const url = titleElement.attr('href');

    urls.push(url);
  });

  return urls;
}

export interface Attachemnt {
  name: string;
  url: string;
}

// TODO: scrape text
interface ScrapedNews {
  title: string;
  id: string;
  date: Date;
  attachments: Attachemnt[];
}

function scrapeNewsPage(page: string): ScrapedNews {
  const $ = cheerio.load(page);

  const title = $('#page-title').text();

  const dateString = $('.field.field-name-field-data-emissione > .field-item > .date-display-single').text();

  // la data passata come ultimo argoemnto serve a prendere i dati che non vengono specificati esplicitamente dalla stringa,
  // ovvero le ore, minuti, secondi, millisecondi
  // usare la data 0 li imposta tutti a 0
  const date = parse(dateString, 'dd/MM/yyyy', new Date(0));

  const id = $('.field.field-name-field-circolare-protocollo > .field-item').text();

  // NOTE: tra .field-item e .file c'è uno spazion e non un >
  // non è fatto a caso, è perchè ci sono altri elementi in mezzo
  const attachments: Attachemnt[] =
    $('.field.field-name-field-circ-all-riservati > .field-item .file > a')
      .toArray()
      .map((tag) => {
        const name = $(tag).text();
        const url = $(tag).attr('href');
        return { name, url };
      });

  return {
    title,
    date,
    id,
    attachments,
  };
}

async function filterNotSent(urls: string[]): Promise<string[]> {
  const promises = urls.map(async (url) => ({
    url,
    sent: await wasSent(url),
  }));

  const objs = await Promise.all(promises);
  return objs
    .filter(({ sent }) => !sent)
    .map(({ url }) => url);
}

export interface News {
  title: string;
  id: string;
  date: Date;
  url: string;
  absoluteUrl: string;
  attachments: Attachemnt[];
}

export async function scrapeNews(newsListPageUrl: string): Promise<News[]> {
  const { data: newsListPage } = await axios.get<string>(newsListPageUrl);

  const urls = scrapeNewsList(newsListPage);
  console.info(`Trovati ${urls.length} link`);

  const newUrls = await filterNotSent(urls);
  console.info(`Trovati ${newUrls.length} nuovi link`);

  const promises: Promise<News>[] = newUrls.map(async (url) => {
    const absoluteUrl = new URL(url, newsListPageUrl).toString();
    const { data: newsPage } = await axios.get<string>(absoluteUrl);
    const scraped = scrapeNewsPage(newsPage);

    return { ...scraped, url, absoluteUrl };
  });

  // Promise.all() fallisce se una promessa qualsiasi fallisce,
  // questo non fallisce mai, al massimo restituice un array vuoto
  const news = (await Promise.allSettled(promises))
    .flatMap(x => {
      if (x.status === 'rejected') {
        console.error('Errore con una circolare', x.reason);
      }

      return x.status === 'fulfilled' ? [x.value] : []
    });
  console.info(`scaricate ${news.length} nuova circolari`);

  // ordina dalla più vecchi alla più nuova
  return news.reverse();
}
