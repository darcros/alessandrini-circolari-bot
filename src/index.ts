import { format } from 'date-fns';
import * as dotenv from 'dotenv';
import { Telegraf } from 'telegraf';
import { scrapeNews } from './news';

import { requireEnv, fixMultinePadding } from './util';

// Carica variabili d'ambiente da .env
dotenv.config();

const TELEGRAM_TOKEN = requireEnv('TELEGRAM_TOKEN');
const BASE_URL = requireEnv('BASE_URL');
const TELEGRAM_CHANNEL_ID = requireEnv('TELEGRAM_CHANNEL_ID');

async function main() {
  console.info('Ottengo elenco circolari...');
  const newsList = await scrapeNews(BASE_URL);
  console.info(`Ottenute ${newsList.length} circolari.`);

  const bot = new Telegraf(TELEGRAM_TOKEN);

  console.info(`Invio ${newsList.length} messaggi`)
  for (const news of newsList) {
    const message = `
      ðŸ“„Circolare numero ${news.id} del ${format(news.date, 'dd/MM/yyyy')}
      
      [${news.title}](${news.absoluteUrl})
    `;

    await bot.telegram.sendMessage(TELEGRAM_CHANNEL_ID, fixMultinePadding(message), { parse_mode: 'MarkdownV2' });
  }
  console.info('finito');
}

main();
