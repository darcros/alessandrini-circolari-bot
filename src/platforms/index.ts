import { Cache } from '../cache';
import { News } from '../news';
import { discordPlatform } from './discord';
import { telegramPlatform } from './telegram';

export interface Bot {
  readonly platformName: string;

  send(news: News[]): Promise<void>;
  send(news: News): Promise<void>;
}

export interface Platform {
  readonly name: string;

  isEnabled: () => boolean;
  initialize: (cache: Cache) => Bot;
}

const platforms = [telegramPlatform, discordPlatform];

export function getEnabledBots(cache: Cache): Bot[] {
  return platforms
    .filter((platform) => {
      const enabled = platform.isEnabled();
      if (!enabled) {
        console.info(`La piattaforma ${platform.name} Ã¨ disabilitata`);
      }

      return enabled;
    })
    .map((platform) => platform.initialize(cache));
}
