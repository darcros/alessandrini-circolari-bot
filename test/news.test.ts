import { readFileSync } from 'fs';
import { ScrapedNews, scrapeNewsList, scrapeNewsPage } from '../src/news';

describe('scraping', () => {
  test('scrapes news list correctly', () => {
    const newsList = readFileSync(
      'test/fixtures/pages/circolari.html'
    ).toString();
    const newsListJson: string[] = JSON.parse(
      readFileSync('test/fixtures/scraped/newsList.json').toString()
    );

    const scraped = scrapeNewsList(newsList);
    expect(scraped).toEqual(newsListJson);
  });

  test('scrapes news correctly', () => {
    const news = readFileSync('test/fixtures/pages/circolare.html').toString();
    const newsJson: ScrapedNews = JSON.parse(
      readFileSync('test/fixtures/scraped/news.json').toString(),

      // converte il campo date da una stringa ad un oggetto Date quando viene letto il file Json
      (key, value) => (key === 'date' ? new Date(value) : value)
    );

    const scraped = scrapeNewsPage(news);
    expect(scraped).toEqual(newsJson);
  });
});
