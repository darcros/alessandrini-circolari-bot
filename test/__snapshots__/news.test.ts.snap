// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scraping scrapes news correctly 1`] = `
Object {
  "attachments": Array [
    Object {
      "name": "2020-2021-c096-orario-di-ricevimento-parenti.pdf",
      "url": "https://www.alessandrinimainardi.edu.it/system/files/circolari/2020/2020-2021-c096-orario-di-ricevimento-parenti.pdf",
    },
  ],
  "date": 2020-10-22T22:00:00.000Z,
  "id": "96",
  "text": "Vittuone, 23/10/2020  

Ai Genitori  
sedi Vittuone e Corbetta 

Oggetto: **Ricevimento parenti per i genitori**

La circolare in oggetto è consultabile in allegato.

Il Dirigente scolastico  
**Prof.ssa Giovanna Ruggeri**",
  "title": " Ricevimento parenti per i genitori",
}
`;

exports[`scraping scrapes news list correctly 1`] = `
Array [
  "/circolari/ricevimento-parenti-i-genitori",
  "/circolari/ricevimento-parenti-i-docenti",
  "/circolari/autocertificazione-assenze-da-malattia-no-covid-19",
  "/circolari/orario-delle-lezioni-dal-19-ottobre-al-24-ottobre",
  "/circolari/ingressi-fuori-orario",
  "/circolari/precisazioni-sull-attuazione-dei-protocolli-di-contenimento-dell-epidemia",
  "/circolari/incontri-e-lezioni-corriere-della-sera",
  "/circolari/orario-delle-lezioni-dal-12-ottobre-al-17-ottobre",
  "/circolari/calendario-esame-di-abilitazione-odontotecnica",
  "/circolari/indicazioni-su-riammissione-studente-scuola",
  "/circolari/giornate-chiusura-istituto-s-2020-2021",
  "/circolari/distribuzione-libretti-giustificazioni",
  "/circolari/orario-delle-lezioni-dal-28-settembre-al-03-ottobre",
  "/circolari/esoneroora-alternativa-alla-religione-cattolica-1",
  "/circolari/gestione-ritardi-o-assenze-nei-collegamenti-da-remoto",
  "/circolari/calendario-scolastico-20202021",
  "/circolari/indicazioni-regione-lombardia-sulla-gestione-dei-casi-covid",
  "/circolari/orario-delle-lezioni-dal-21-al-26-settembre",
  "/circolari/gestione-assenze-studenti",
  "/circolari/quadro-orario-dal-15-settembre-al-19-settembre-classi-prime",
]
`;
